teps/XXXX-step-workspaces.md

```diff
@@ -0,0 +1,254 @@
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+    - [Story 4](#story-4)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Use an explicit volumeMount instead](#use-an-explicit-volumemount-instead)
+    - [Advantages](#advantages)
+    - [Drawbacks](#drawbacks-1)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+## Motivation
+
+### Goals
+
+- Provide a mechanism to limit the exposure of sensitive workspaces to only those Steps in a Task
+that actually require access to them.
+- Provide explicit access to Task workspaces from sidecars without using `volumeMounts`.
+
+### Non-Goals
+
+- Validation of the contents of a WorkspaceBinding
+- Validation of the configuration of a WorkspaceBinding
+- Overriding configuration of a WorkspaceBinding
+
+## Requirements
+
+- A Task author should be able to limit access to a `Workspace` to only those `Steps` that actually
+require the contents of that `Workspace`. By doing so they should be able to limit the running code
+that has access to those contents as well.
+- A Task author should be able to use a Workspace from a `Sidecar`.
+
+## Proposal
+
+### Add `workspaces` to `Steps`
+
+1. Add a `workspaces` list to `Steps`.
+2. Allow `workspaces` from the `Task` to be explicitly named in that list like this:
+
+    ```yaml
+    workspaces:
+    - name: my-sensitive-workspace
+    steps:
+    - name: foo
+      workspaces:
+      - name: my-sensitive-workspace
+    ```
+3. A `workspace` named in a `Step` is not given to other `Steps` unless they too have the
+`workspace` in their own `workspaces` list.
+
+Example YAML:
+
+```yaml
+# task spec
+spec:
+  workspaces:
+  - name: git-ssh-credentials
+    mountPath: /root/.ssh
+  steps:
+  - name: clone-repo
+    image: alpine/git:v2.26.2
+    workspaces:
+    - name: git-ssh-credentials
+    script: |
+      git clone $(params.repo-url) /workspace/source
+  - name: run-unit-tests
+    script: |
+      cd /workspace/source
+      go test ./...
+```
+
+In the above example only the `clone-repo` `Step` will receive access to the `git-ssh-credentials`
+`Workspace`. The `run-unit-tests` `Step` will not receive access to the Workspace volume. Importantly
+this also means that the user-supplied code does not have access to the credential files. Compromising
+the code for the unit tests does not also compromise the SSH credentials.
+
+### Add `workspaces` to `Sidecars`
+
+1. Add a `workspaces` list to `Sidecars`.
+2. Allow `workspaces` from the `Task` to be explicitly named in that list like this:
+
+    ```yaml
+    workspaces:
+    - name: my-workspace
+    sidecars:
+    - name: watch-workspace
+      workspaces:
+      - name: my-workspace
+    ```
```

Comment:
is this in TaskRun or Pipeline or both? im mostly wondering b/c to support this for sidecars i think we'd have to include this in "pipelinetasks" in a Pipeline as well (which up until now have no "sidecars" section i dont think?)

Reply:
ah kk, i thought this was at runtime. nevermind me